src/control/cmd/agent/infocache.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/daos-stack/daos/src/control/lib/atm"
+	"github.com/daos-stack/daos/src/control/lib/netdetect"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+const defaultNumaNode = 0
+
+type attachInfoCache struct {
+	log logging.Logger
+	// is caching enabled?
+	enabled atm.Bool
+	// is the cache initialized?
+	initialized atm.Bool
+	// maps NUMA affinity and device index to a response
+	resmgmtpb map[int]map[int][]byte
+	// maps NUMA affinity to a device index
+	devIdx map[int]int
+	mutex  sync.Mutex
+}
+
+// loadBalance is a simple round-robin load balancing scheme
+// to assign network interface adapters to clients
+// on the same NUMA node that have multiple adapters
+// to choose from.  Returns the index of the device to use.
+func (aic *attachInfoCache) loadBalance(numaNode int) int {
+	aic.mutex.Lock()
+	numDevs := len(aic.resmgmtpb[numaNode])
+	// keep the deviceIndex in range of the current resmgmtpb map
+	deviceIndex := aic.devIdx[numaNode] % numDevs
+	aic.devIdx[numaNode] = (deviceIndex + 1) % numDevs
+	aic.mutex.Unlock()
+	return deviceIndex
+}
+
+func (aic *attachInfoCache) getResponse(numaNode int) ([]byte, error) {
+	deviceIndex := aic.loadBalance(numaNode)
+	aic.mutex.Lock()
+	resmgmtpb, ok := aic.resmgmtpb[numaNode][deviceIndex]
+	aic.mutex.Unlock()
+	if !ok {
+		return nil, errors.Errorf("GetAttachInfo entry for numaNode %d device index %d did not exist", numaNode, deviceIndex)
+	}
+	aic.log.Debugf("Retrieved response for NUMA %d with device index %d\n", numaNode, deviceIndex)
+	return resmgmtpb, nil
+}
+
+// initResponseCache generates a unique dRPC response corresponding to each device specified
+// in the scanResults.  The responses are differentiated based on the network device NUMA affinity.
+func (aic *attachInfoCache) initResponseCache(resp *mgmtpb.GetAttachInfoResp, scanResults []netdetect.FabricScan) error {
+	aic.mutex.Lock()
+	defer aic.mutex.Unlock()
+
+	if len(scanResults) == 0 {
+		return errors.Errorf("No devices found in the scanResults")
+	}
+
+	// Make a new map each time the cache is initialized
+	aic.resmgmtpb = make(map[int]map[int][]byte)
+
+	// Make a new map just once.
+	// Preserve any previous device index map in order to maintain ability to load balance
+	if len(aic.devIdx) == 0 {
+		aic.devIdx = make(map[int]int)
+	}
+
+	netdetect.SetLogger(aic.log)
+
+	for _, fs := range scanResults {
+		if fs.DeviceName == "lo" {
+			continue
+		}
+		resp.Interface = fs.DeviceName
+		// by default, the domain is the deviceName
+		resp.Domain = fs.DeviceName
+		if strings.HasPrefix(resp.Provider, "ofi+verbs") {
+			deviceAlias, err := netdetect.GetDeviceAlias(resp.Interface)
+			if err != nil {
+				aic.log.Debugf("non-fatal error: %v. unable to determine OFI_DOMAIN for %s", err, resp.Interface)
+			} else {
+				resp.Domain = deviceAlias
+				aic.log.Debugf("OFI_DOMAIN has been detected as: %s", resp.Domain)
+			}
+		}
+		numa := int(fs.NUMANode)
+
+		resmgmtpb, err := proto.Marshal(resp)
```

Comment:
I have a hard time defending why I should cache anything other than the final result, when any other answer requires more compute cycles.  I'll keep it as-is, and if the team decides we would rather have slower performance in the cache in favor of less complexity, I can go ahead and change it.

Reply:
Fair, although marshaling a cached response to pass back to the client is still pretty cheap compared to the full sequence. If leaving it as-is, it would be nice to have the variable names indicate that it's marshaled.

Reply:
Done.